File: Grind 75 in Java/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();

        dfs(target, 0, 0, new ArrayList<Integer>(), res, candidates);
        return res;
    }

    private void dfs(int t, int sum, int i, List<Integer> tmp, List<List<Integer>> res, int[] cnd) {
        if (sum > t) return;
        if (sum == t) {
            res.add(List.copyOf(tmp));
            return;
        }

        for (int j = i; j < cnd.length; j++) {
            int n = cnd[j];
            tmp.addLast(n);
            dfs(t, sum + n, j, tmp, res, cnd);
            tmp.removeLast();
        }
    }
}

File: leetcode in kotlin/0039-combination-sum.kt
class Solution {
    fun combinationSum(candidates: IntArray, target: Int): List<List<Int>> {
        val res = ArrayList<ArrayList<Int>>()
        val combi = ArrayList<Int>()
        dfs(0, target, combi, res, candidates)
        return res
    }
    //fun for backtracking
    private fun dfs(index: Int, target: Int, temp: ArrayList<Int>, res: ArrayList<ArrayList<Int>>, cand: IntArray){
        if(target == 0){
            res.add(ArrayList(temp))
            return
        }
        if(index >= cand.size || target < 0)
            return
        temp.add(cand[index])
        dfs(index, target-cand[index], temp, res, cand)
        temp.removeAt(temp.size-1)
        dfs(index+1, target, temp, res, cand)
    }
}
